### ISSUE TITLE
Reorganise dmipy_sim into layered packages (flat 32-module namespace -> substrate/engine/replay/...)

### ISSUE THREAD

**Issue by rutgerfick** [dmrai-lab/dmipy-sim#88]
## Why

`dmipy_sim/` is 19,845 lines across 32 flat modules. `geometries.py` alone is 3,548 lines
and 12 classes (`PackedCylinders` is 787 of them). Tests are 88 files / 14,483 lines, flat
apart from `physics/` and `validation/`, and their names do not pair with the modules they
cover — so "what do I re-run after editing X" is a guess, which is expensive when CI is
~80 min.

## The finding that makes this tractable

**The layering already exists and is acyclic.** The internal import graph sorts into five
clean levels with no cycles:

| layer | modules |
|---|---|
| 0 | `_boundary`, `constants`, `gpu`, `_gpu_config`, `gaunt`, `noise`, `susceptibility_field` |
| 1 | `geometries`, `physics`, `waveforms`, `rf`, `compression`, `trajectories`, `susceptibility` |
| 2 | `core`, `bloch`, `mesh`, `curved_tube`, `mt`, `mt_walk`, `replay`, `mesh_shapes`, `viz`, `pedagogy` |
| 3 | `bank`, `phantom`, `pulse_sequence`, `mesh_axon` |
| 4 | `bank_cactus`, `mesh_bundle`, `sh_convolution` |

Nothing needs re-architecting. The structure is latent and correct; expressing it as
directories is a **rename**, not a redesign. That is what caps the risk.

## Target layout

```
dmipy_sim/
  __init__.py           # public API — unchanged, this is the contract
  constants.py
  substrate/   _boundary, base, analytic, packed, myelin, packing, curved_tube, mesh, mesh_shapes
  engine/      physics, core, bloch, pulse_sequence, mt, mt_walk, gpu, _gpu_config
  acquisition/ waveforms, rf, noise
  replay/      trajectories, compression, replay, bank, phantom, sh_convolution, gaunt, builders/
  fields/      susceptibility, susceptibility_field
  viz/         viz, pedagogy
```

Two placements are load-bearing rather than aesthetic: `pulse_sequence` goes in `engine/`
beside `bloch` (it is bloch-specific; putting it in `acquisition/` creates a package-level
cycle), and the one remaining back-edge `core -> trajectories` is *already* a deferred
function-level import, so it does not bind at module load.

Tests mirror the package, one file per module, so a test path is derivable from the module
it covers. `tests/validation/` and `tests/physics/` stay as deliberate exceptions — they
assert physics across modules and should not be forced into a 1:1 mapping.

## Constraint: downstream imports submodules

`dmipy-fit` and `dmipy-design` import submodule paths, not just the package:
`dmipy_sim.replay` (12x), `constants` (8x), `waveforms` (5x), `sh_convolution` (5x),
`gaunt` (4x), `geometries` (3x), `compression` (3x), `rf`, `bank`, `pulse_sequence`.

Every move therefore ships a one-line re-export shim at the old flat path. No downstream
change is required by this work.

**This is already overdue:** three downstream imports are broken today —
`dmipy_sim.sequences`, `dmipy_sim.canonical` and `dmipy_sim.substrate` are imported by
dmipy-fit and do not exist. The flat namespace has already rotted silently, and nothing
caught it.

## Steps (one PR each, each independently green)

- [ ] **1. `geometries.py` -> `substrate/`** — the 3,548-line file becomes `base` (~130) /
      `analytic` (~1400) / `packed` (~1300) / `myelin` (~650) / `packing`, with `_boundary`
      moved in beside its six consumers. Highest value, self-contained.
- [ ] **2. `tests/substrate/`** mirroring it; move the matching tests.
- [ ] **3. Public-API surface test** + fix the three rotted downstream imports.
- [ ] **4. `engine/`** (physics, core, bloch, pulse_sequence, mt, mt_walk, gpu).
- [ ] **5. `replay/`** (+ `builders/`) and **`acquisition/`**, **`fields/`**, **`viz/`**.

Do not batch these. CI is ~80 min, so a bad batch costs a full cycle to bisect.

## Scope note

This is ~34k lines of pure motion with **no behaviour change**. The payoff is ergonomic,
not correctness — nothing here fixes a bug. Every PR must land with the fast tier green and
no test edits beyond moves and import updates.

Full analysis: `docs/REORG_PROPOSAL.md`.

**Comment by rutgerfick** [comment]
## Correction: the premise of this issue was partly wrong

I globbed `dmipy_sim/*.py` and concluded "32 flat modules". That glob does not see
directories, so I missed that **`dmipy_sim` already has four subpackages** with a
deliberate convention:

| package | docstring | files | lines |
|---|---|---|---|
| `substrate/` | "The physical forward-truth substrate (sim-owned)" | 4 | 3,798 |
| `sequences/` | "Physical acquisition sequences (sim-owned forward-signal definition)" | 5 | 1,345 |
| `math/` | "Self-contained math utilities for dmipy-sim (the forward truth)" | 3 | 177 |
| `io/` | substrate loaders (CACTUS, Winther, mesh) | 4 | 469 |

Three consequences, all of which change the plan:

**1. Two of the three "broken downstream imports" were not broken.** `dmipy_sim.sequences`
and `dmipy_sim.substrate` resolve fine — they are packages, and my check tested for
`substrate.py` as a *file*. Only `dmipy_sim.canonical` is genuinely missing. I have
withdrawn that claim; the "namespace has already rotted" framing was overstated on a
2-of-3 error.

**2. The proposed names collide with existing, different meanings.** `substrate/` here
means the *biophysical tissue specification* (parameters, validators, constants) — not
geometry. `sequences/` already occupies what I called `acquisition/`. Reusing either name
for geometry would be actively worse than the flat layout.

**3. There is already a convention to extend rather than invent.** The existing packages
are organised by *ownership and role* ("sim-owned forward truth"), not by call layer.

## Revised step 1

Geometry classes go to **`geometry/`**, not `substrate/`:

```
dmipy_sim/geometry/
  _boundary.py     # shared boundary rules, beside its consumers
  base.py          # Geometry ABC, FreeDiffusion, Box1D, _rotation_to_z   ~160 l
  analytic.py      # Sphere, Cylinder, Ellipsoid, PermeableSlab1D, Shell  ~1400 l
  packed.py        # PackedCylinders, PackedSpheres                       ~1300 l
  myelin.py        # MyelinatedCylinder, PackedMyelinatedCylinders         ~650 l
  packing.py       # pack_cylinders / pack_spheres / pack_myelinated
  curved_tube.py  mesh.py  mesh_shapes.py
```

The split itself is confirmed clean by AST analysis: the **only** cross-file dependency
among the 12 classes is the `Geometry` ABC, plus one shared helper (`_rotation_to_z`, used
by 4 classes). One import line per new file.

The remaining steps stand, with `acquisition/` dropped (it is `sequences/`) and the
engine/replay/fields grouping re-checked against the existing packages before each PR.

Apologies for the noise — the measurement was wrong, so the plan built on it was too.

**Comment by rutgerfick** [comment]
## Step 1b: collapse `PackedCylinders` / `PackedSpheres` onto a shared radial traversal

Raised while reviewing step 1: `geometry/packed.py` is 1,158 lines for two classes, which is
more than the physics needs. Measured breakdown:

- **only 55% is executable code** (636 lines; the rest is 222 docstring, 159 comment, 141 blank)
- **~69% of that code is three methods**, written once per class:

| method | PackedCylinders | PackedSpheres |
|---|---|---|
| `reflect` | 135 | 89 |
| `reflect_with_log_weight` | 128 | 90 |
| `permeate` | 214 | 134 |

Those pairs are **the same algorithm**. Normalising only the 2-D/3-D naming
(`xy_final`↔`r_out`, `d_hat_xy`↔`d_hat`, `centers_2d`↔`centers`, …) they are:

```
reflect                  64% identical
reflect_with_log_weight  70% identical
permeate                 73% identical
```

`_boundary` extracted the boundary *rules*. What is still duplicated is the *traversal
scaffold*: rotate into the object frame → minimum-image offsets to all N objects → pick the
nearest valid root → snap to the surface → decide transmit/reflect → reassemble. Identical
logic, spelled once for a 2-vector and once for a 3-vector.

Diffing the residual 27–36% shows it is three things:

1. **Essential** (~10 lines) — the axial split (`r_c[:2]`, `step_z`, `z_final`) and the
   rotation frame. A cylinder has an unrestricted axis; a sphere does not.
2. **Docstrings** (~9 lines) — not code.
3. **Accidental** (~15 lines) — the carried-`side=` API added in #87 to `PackedCylinders`
   but not `PackedSpheres`. That asymmetry is the single largest chunk of the difference,
   and it is drift, not physics.

### The pattern is already in this codebase

`PermeableShell` handles both shapes with one implementation:

```python
def _radial(self, r):
    """Radial vector used for the membranes (full r for sphere, perpendicular for cyl)."""
    if self.kind == 'sphere':
        return r
    return r - jnp.dot(r, self._axis) * self._axis
```

Express the packed classes through the same projection and the three boundary methods
collapse to one each — roughly **300 lines of duplicated code removed**, and the `side=`
asymmetry disappears by construction rather than needing to be ported.

### Why this is now feasible

Unlike the pure-motion steps, this changes executed code. It is checkable because #87 left a
real net: 26 compartment-integrity tests probing all seven geometries adversarially
(bisect to the wall, step exactly onto it), 19 direct primitive unit tests, and a
crossing-rate suite validated against the closed two-compartment exchange law. A collapse
that breaks either packed geometry fails loudly.

**Sequence:** after step 2 (tests move), before step 4 (`engine/`). It must be its own PR
with its own CI cycle — not batched with a motion-only step, because a regression here would
be indistinguishable from a bad move.

**Comment by rutgerfick** [comment]
## Step 1b, rewritten: one wall-interaction API (supersedes the packed-dedup framing above)

The comment above proposed collapsing `PackedCylinders`/`PackedSpheres` onto a shared radial
traversal. That was too narrow. The real duplication is not between the two packed classes —
it is between `reflect`, `reflect_with_log_weight` and `permeate`, which are **the same
function at different argument values**, repeated across every geometry.

### Evidence

**They are parameterisations of one another.** Measured on `PackedCylinders`, 20,000 walkers:
`reflect` and `permeate(kappa=0)` are **bit-identical** on extra-axonal walkers (0 differing
of 11,332) — same algorithm, not merely similar.

**The three buy nothing.** 40,000 walkers, one step:

| | cost |
|---|---|
| `reflect` | 0.11 ms |
| `permeate(kappa=0)` | 0.11 ms |
| `permeate(kappa>0)` | 0.41 ms |

XLA constant-folds `kappa=0` and eliminates the dead transmit branch, so the separate
`reflect` is not an optimisation. (The `kappa>0` figure is real work — bidirectional root
selection and the transmit draw — not overhead.)

**The cost of keeping them apart:** 36 implementations (18 `reflect`, 10
`reflect_with_log_weight`, 8 `permeate`), plus 23 lines of engine dispatch whose only job is
choosing between them. **No downstream caller** touches any of the three, so the API is
free to change.

### And the split is actively generating bugs

`PackedCylinders.reflect` **expels 100% of intra-axonal walkers** at kappa = 0. Measured,
4,000 walkers seeded inside cylinders, 600 steps:

```
reflect         intra walkers that left their cylinder:  4000/4000 (100.00%)
permeate(k=0)   intra walkers that left their cylinder:     0/4000 (  0.00%)
```

At kappa = 0 nothing passes, so a walker reflects regardless of which side of the wall it is
on. `reflect` is simply wrong on one side — the same algorithm, silently divergent.

**This is reachable by default.** The engine selects the path with
`has_permeability = permeability is not None`:

```
PackedCylinders(centers=..., radii=..., L=...)      -> permeability=None -> uses REFLECT (broken for intra)
PackedCylinders(..., permeability=0.0)              -> uses permeate     -> correct
```

So building an impermeable packed substrate the natural way — by omitting `permeability` —
gets the broken path, and you only get correct intra confinement by passing `permeability=0.0`,
which reads like a no-op. Same for `PackedSpheres`. This is the fourth instance of the #86
defect class (a boundary path that does not conserve compartment), and it is in the method
most users reach first.

### Target

```python
def interact(self, r, step, *, kappa_over_D=0.0, rho_over_D=0.0,
             key=None, side=None) -> WallHit:
    """One wall interaction: reflect, or cross if the membrane grants it."""

class WallHit(NamedTuple):      # a pytree -- JAX handles it natively
    r:       jnp.ndarray        # new position
    dlog_w:  jnp.ndarray        # surface log-weight increment (0 when rho_over_D = 0)
    crossed: jnp.ndarray        # bool: a crossing was granted
    illegal: jnp.ndarray        # bool: the compartment sentinel fired
```

This also fixes the ragged returns (`permeate` currently returns a 2-tuple or a 4-tuple
depending on whether `side` was passed).

**One thing the three-method design was doing for us:** capability advertisement. Today
"can this geometry permeate?" is answered by `hasattr(geometry, 'permeate')` —
`FreeDiffusion` has only `reflect`, `Box1D` has no `permeate`. With one method that becomes
an explicit `supports_permeability` flag plus a clear error when kappa > 0 reaches a geometry
that cannot cross. That is better than an `AttributeError`-driven fallback, but it is a real
piece of behaviour to port rather than drop.

### Consequences

- The packed-class dedup falls out for free: implementing one method instead of three
  collapses ~300 lines of scaffold as a side effect rather than as the goal.
- **Not behaviour-preserving, deliberately:** extra-axonal results stay bit-identical
  (verified); intra-axonal results under `reflect` change from wrong to correct.
- Verification is the #87 suite: 26 compartment-integrity tests probing all seven geometries
  adversarially, 19 primitive unit tests, and crossing rates against the closed
  two-compartment exchange law. The intra-confinement regression above becomes a test.

**Comment by rutgerfick** [comment]
## Status and what is left

**Step 1 — `geometries.py` → `geometry/` package.** Merged (#89, `c841902`). The 3,548-line
file became `base` / `analytic` / `packed` / `myelin` / `packing`, with `_boundary` moved in
beside its consumers and `curved_tube` / `mesh` / `mesh_shapes` alongside. Five compat shims
keep `dmipy_sim.geometries` and friends working, so no downstream change was required.

**Step 1b — one wall interaction per geometry.** Open as #90 (+1,055 / −1,237, 13 files),
awaiting CI. This grew well beyond the original "collapse the two packed classes" framing:
`reflect`, `reflect_with_log_weight` and `permeate` were the same function at different
argument values, and keeping them apart had produced four separate bugs. See the PR.

### Remaining

- [ ] **2. `tests/geometry/`** mirroring the package; move the matching test files so a test
      path is derivable from the module it covers.
- [ ] **3. Public-API surface test**, plus the one genuinely broken downstream import
      (`dmipy_sim.canonical`, imported by dmipy-fit, does not exist).
- [ ] **4. `engine/`** — `physics`, `core`, `bloch`, `pulse_sequence`, `mt`, `mt_walk`, `gpu`.
- [ ] **5. `replay/`** (+ `builders/`), `fields/`, `viz/`. No `acquisition/` — `sequences/`
      already occupies that role.

### New work this surfaced, worth its own issues

- [ ] **Engine still dispatches on `hasattr`.** `core.py` / `physics.py` / `bloch.py` carry
      ~23 lines choosing between `reflect` / `reflect_with_log_weight` / `permeate`. Now that
      `Geometry.interact` exists with explicit `supports_permeability` / `carries_side`
      capabilities, that dispatch should collapse to one call.
- [ ] **`Box1D.reflect_with_log_weight` is still its own 38-line implementation**, because
      Box1D has no `permeate` to fold into. Either give it one (a reflecting box is the
      kappa = 0 case of a slab) or route it through `interact`.
- [ ] **`bank_cactus.py` holds a 512-line fork of `mesh_bundle.py`'s `BoxedMesh`**, 861 lines
      divergent from the tracked copy, and it imports `.io.mcdc` which does not exist. It is
      untracked, so it received none of the boundary fixes in #87 or #90.
- [ ] **`BoxedMesh._box` carries its own compartment veto**, whose docstring already says it
      is "a mitigation, not the design fix" — box faces should take part in the same bounce
      loop as the mesh triangles.
- [ ] **CI takes ~80 minutes for the "fast" tier**, which `CLAUDE.md` documents as "~1 min on
      CPU". That claim is badly stale and the tier needs re-triaging; a per-test timing run is
      in progress.

### Method note

The impact table added in #90 (`tests/test_wall_impacts.py`) is the tool that made this
tractable: fire one walker at one wall from a known offset x direction x step length, no
walk, no statistics. It caught a multi-bounce regression in milliseconds that the MC suite
took 19.5 minutes to report, and it names the failing case rather than saying "a walker
escaped". Any future boundary change should extend that table first.

**Comment by rutgerfick** [comment]
Status against the step list: (1) done as `dmipy_sim/geometry/` (base, analytic, packed, myelin, packing, _boundary, _grid, curved_tube, mesh, mesh_shapes) with warning shims at the old flat paths; (2) done in #117 (`tests/geometry/`, 35 modules, pure motion); (3) already covered by `tests/test_api_surface.py::test_submodule_imports` over every submodule path dmipy-fit / dmipy-design import, incl. `dmipy_sim.sequences` and `dmipy_sim.substrate` -- `dmipy_sim.canonical` is a private-repo module that dmipy-fit skip-guards, not a public rot; (4) `engine/` is next, one PR; (5) after it. The pre-existing `substrate/` package (the Substrate dataclass and constants) keeps its name, so the geometry package is `geometry/`, not `substrate/`.

**Comment by rutgerfick** [comment]
All five steps are on main: (1) `geometry/`; (2) `tests/geometry/` #117; (3) the API-surface import list in tests/test_api_surface.py now carries every old flat path AND every packaged path (58 module imports); (4) `engine/` #118; (5) `replay/` + `replay/builders/` #119 and `acquisition/`, `fields/`, `viz/` #120. Each PR was pure motion validated three ways: the package imports with DeprecationWarning promoted to an error (so nothing inside dmipy_sim goes through a shim), the collected test count is unchanged apart from the new import tests, and the full fast tier is green on the GPU. Old flat paths are one-line shims warning DeprecationWarning, except `dmipy_sim.replay` and `dmipy_sim.viz`, which are the packages themselves re-exporting the same-named module, so downstream imports of them are not deprecated. Final layout: `geometry/ engine/ replay/(builders/) acquisition/ fields/ viz/ sequences/ substrate/ io/ math/` plus the level-0 modules `constants`, `compartments`, `persistent_walk` at the top. Leaving the issue open for you to close; the shims are the one follow-up (remove next release).

**Comment by rutgerfick** [comment]
Shims removed (PR #121): the package is the layered layout only, no flat-path modules remain. `dmipy_sim.replay` and `dmipy_sim.viz` are the packages re-exporting their same-named module. Downstream (dmipy-fit / dmipy-design) will need their imports moved to the packaged paths when they next pin dmipy-sim; the API-surface test lists every packaged module path. Nothing left open on this issue.
